### ISSUE TITLE
Is SoftmaxOutput actually using logloss as the error function?

### ISSUE THREAD

**Issue by SkidanovAlex** [dmlc/mxnet#1199]
In the docs for SoftmaxOutput it says:

> backprop with logloss.

However, when I look at `SoftmaxOutputOp`, the `Backward` functions calls to `SoftmaxGrad` from mshadow, which is in turn implemented like this:

```
template<typename DType>
inline void SoftmaxGrad(Tensor<cpu, 2, DType> dst,
                        const Tensor<cpu, 2, DType> &src,
                        const Tensor<cpu, 1, DType> &label) {
  for (index_t y = 0; y < dst.size(0); ++y) { 
    const index_t k = static_cast<int>(label[y]);
    for (index_t x = 0; x < dst.size(1); ++x) {
      if (x == k) {
        dst[y][k] = src[y][k] - 1.0f;
      } else {
        dst[y][x] = src[y][x];
      }
    }
  } 
} 
```

It does not appear to be logloss to me, unless I am reading it wrong. Is it computed in some other place?

**Comment by jermainewang** [comment]
I think the doc means `SoftmaxOutputOp` computes logloss so the `SoftmaxGrad` computes the gradient of logloss which is what you saw in the code.

**Comment by SkidanovAlex** [comment]
@jermainewang that was my first guess, but between `SoftmaxOutputOp::Forward`, `SoftmaxOutputOp::Backward` from mxnet and `Softmax` and `SoftmaxGrad` from mshadow I can't find any place where logloss would be computed.

**Comment by tornadomeet** [comment]
@SkidanovAlex , there is no real calc of log-loss in mxnet now, because we need not loss for training(but only need the grad of it), you can add the calc of log-loss in the python side.

**Comment by SkidanovAlex** [comment]
@tornadomeet I see. I got confused because I didn't realize that the derivative of softmax + logloss is just p_i - y_i.

In my case my last layer is Softmax, but my loss function needs to have an extra term in it. Is my understanding right that I will need to implement a custom output layer for this usecase?

**Comment by tornadomeet** [comment]
@SkidanovAlex yes, in your case, you shoud use SoftmaxActivation instead of SoftmaxOutput, and then write your own loss layer~

**Comment by salemmohammed** [comment]
@tornadomeet  how to add log loss in python side ?

**Comment by tornadomeet** [comment]
ref `CrossEntropy` in `python/mxnet/metric.py`
